File: examples/mpi/c/allgatherv.c
/****************************************************************
 ****************************************************************
 ****
 ****
 **** This program file is part of the book and course
 **** "Parallel Computing"
 **** by Victor Eijkhout, copyright 2013-6
 ****
 **** allgatherv.c : vector allgather, task t contributes t+1 data
 ****
 ****************************************************************
 ****************************************************************
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "mpi.h"

int main(int argc,char **argv) {
  
#include "globalinit.c"
  
  int my_count = procno+1;
  int *my_array = (int*) malloc(my_count*sizeof(int));
  for (int i=0; i<my_count; i++)
    my_array[i] = procno;
  int *recv_counts = (int*) malloc(nprocs*sizeof(int));
  int *recv_displs = (int*) malloc(nprocs*sizeof(int));

  MPI_Allgather
    ( &my_count,1,MPI_INT,
      recv_counts,1,MPI_INT, comm );
  int accumulate = 0;
  for (int i=0; i<nprocs; i++) {
    recv_displs[i] = accumulate; accumulate += recv_counts[i]; }
  int *global_array = (int*) malloc(accumulate*sizeof(int));
  MPI_Allgatherv
    ( my_array,procno+1,MPI_INT,
      global_array,recv_counts,recv_displs,MPI_INT, comm );
  
  if (procno==0) {
    for (int p=0; p<nprocs; p++)
      if (recv_counts[p]!=p+1)
	printf("count[%d] should be %d, not %d\n",
	       p,p+1,recv_counts[p]);
    int c = 0;
    for (int p=0; p<nprocs; p++)
      for (int q=0; q<=p; q++)
	if (global_array[c++]!=p)
	  printf("p=%d, q=%d should be %d, not %d\n",
		 p,q,p,global_array[c-1]);
  }

  MPI_Finalize();
  return 0;
}

File: examples/mpi/p/allgatherv.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
#### "Parallel Computing"
#### by Victor Eijkhout, copyright 2013-6
####
#### allgatherv.py : vector allgather, task t contributes t+1 data
####
################################################################
################################################################

import numpy as np
import random # random.randint(1,N), random.random()
import sys
from mpi4py import MPI

comm = MPI.COMM_WORLD
procid = comm.Get_rank()
nprocs = comm.Get_size()
if nprocs<2:
    print("C'mon, get real....")
    sys.exit(1)

##codesnippet bufallocp
mycount = procid+1
my_array = np.empty(mycount,dtype=np.float64)
##codesnippet end

for i in range(mycount):
    my_array[i] = procid
recv_counts = np.empty(nprocs,dtype=int)
recv_displs = np.empty(nprocs,dtype=int)

##codesnippet allgathervp
my_count = np.empty(1,dtype=int)
my_count[0] = mycount
comm.Allgather( my_count,recv_counts )

accumulate = 0
for p in range(nprocs):
    recv_displs[p] = accumulate; accumulate += recv_counts[p]
global_array = np.empty(accumulate,dtype=np.float64)
comm.Allgatherv( my_array, [global_array,recv_counts,recv_displs,MPI.DOUBLE] )
##codesnippet end

# other syntax:
# comm.Allgatherv( [my_array,mycount,0,MPI.DOUBLE], [global_array,recv_counts,recv_displs,MPI.DOUBLE] )

if procid==0:
    #print(procid,global_array)
    for p in range(nprocs):
        if recv_counts[p]!=p+1:
            print( "recv count[%d] should be %d, not %d" \
                % (p,p+1,recv_counts[p]) )
    c = 0
    for p in range(nprocs):
        for q in range(p+1):
            if global_array[c]!=p:
                print( "p=%d, q=%d should be %d, not %d" \
                    % (p,q,p,global_array[c]) )
            c += 1
    print "finished"
    
